src/main/java/net/dv8tion/jda/internal/JDAImpl.java

```diff
@@ -754,6 +752,20 @@ public synchronized void shutdown()
             client.shutdown();
         }
     }
+    
+    @Override
+    public void awaitShutdown() throws InterruptedException
+    {
+        threadConfig.awaitTermination();
+        WebSocketClient client = getClient();
+        if (client != null)
+        {
+            while (!client.isShutdownEventFired())
```

Comment:
Why aren't we checking for status instead? If its because `shutdownInternals` sets it, we could change it to only be set by the websocket instead. Maybe make a `JDAImpl#signalShutdown` which updates the status and fires the event.

Reply:
The original intent was to sync the return on firing of the ShutdownEvent, so that flag was used wherever the event would have been fired. `shutdownInternals` doesn't directly fire the event, but does set the status; websocket `handleDisconnect` does internals but fires the event; websocket `reconnect`, `queueReconnect`, and constructor both set status and fire.

It seems all of these cases at least set the status flag somewhere along the track, so in hindsight I agree that should be a status check.